### PR TITLE
Linear regression and logistic regression

### PR DESCRIPTION
**Pull request by sugarspectre**

### CODE REVIEW

File: src/model/linear-model/linear-regression.ts [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+   */
+  fitIntercept?: boolean,
+  /**
+   * This parameter is ignored when fit_intercept is set to False. If True, the regressors X will be normalized before regression by subtracting the mean and dividing by the l2-norm.
+   */
+  normalize?: boolean,
+  /**
+   * Optimizer types
+   */
+   optimizerType?: OptimizerType,
```
Comment (FeelyChau) [inline src/model/linear-model/linear-regression.ts:23]: indent

File: src/model/linear-model/linear-regression.ts [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+   /**
+    * Optimizer properties
+    */
+   optimizerProps?: OptimizerProps
```
Comment (FeelyChau) [inline src/model/linear-model/linear-regression.ts:27]: ditto

**Review by FeelyChau — CHANGES_REQUESTED** [review]

File: package.json [hunk 3]
```diff
@@ -4,10 +4,13 @@
...
   "main": "dist/index.js",
   "dependencies": {
+    "@babel/preset-env": "^7.15.4",
     "@tensorflow/tfjs-core": "^3.8.0",
     "@tensorflow/tfjs-layers": "^3.8.0",
+    "babel-loader": "^8.2.2",
     "jimp": "^0.16.1",
-    "seedrandom": "^3.0.5"
+    "seedrandom": "^3.0.5",
+    "webpack-visualizer-plugin": "^0.1.11"
```
Comment (FeelyChau) [inline package.json:13]: Move to `devDependencies'.

File: src/preprocess/encoder.ts [hunk 4]
```diff
@@ -49,7 +49,13 @@ export class OneHotEncoder {
    */
   public async init(x: Tensor | number[] | string[]): Promise<void> {
     const { values } = unique(x);
-    this.categories = values;
+    if (values.dtype == 'int32' || values.dtype == 'float32') {
```
Comment (FeelyChau) [inline src/preprocess/encoder.ts:52]:
```suggestion
    if (values.dtype === 'int32' || values.dtype === 'float32') {
```

File: src/preprocess/encoder.ts [hunk 5]
```diff
@@ -49,7 +49,13 @@ export class OneHotEncoder {
...
+      this.categories = topk(values, values.shape[0], false).values;
+    } else if (values.dtype == 'bool') {
```
Comment (FeelyChau) [inline src/preprocess/encoder.ts:54]:
```suggestion
    } else if (values.dtype === 'bool') {
```

File: src/preprocess/encoder.ts [hunk 6]
```diff
@@ -92,11 +98,11 @@ export class OneHotEncoder {
     }
     const nCate = this.categories.shape[0];
     const codeSize = this.drop === 'first' ? nCate - 1 : this.drop === 'binary-only' && nCate === 2 ? 1 : nCate;
-    const shapeCorrect = codeSize > 1 ? checkShape(x, [ -1, codeSize ]) : checkShape(x, [ -1 ]);
+    const shapeCorrect = codeSize > 1 ? checkShape(x, [ -1, codeSize ]) : (checkShape(x, [ -1 ]) || (checkShape(x, [ -1, 1 ])));
```
Comment (FeelyChau) [inline src/preprocess/encoder.ts:101]:
```suggestion
    const shapeCorrect = codeSize > 1 ? checkShape(x, [ -1, codeSize ]) : (checkShape(x, [ -1 ]) || checkShape(x, [ -1, 1 ]));
```

File: browser-examples/logistic-regression.html [hunk 7]
```diff
@@ -0,0 +1,117 @@
+<html>
```
Comment (yorkie) [inline browser-examples/logistic-regression.html:1]: examples/browser or examples/web

File: src/utils/optimizer-types.ts [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+ * @param optimizerProps optimizer properties
+ */
+export const getOptimizer = (optimizerTypes: OptimizerType, optimizerProps: OptimizerProps): Optimizer => {
+  switch (optimizerTypes) {
+  case 'sgd': {
+    const props = (optimizerProps as SGDProps);
+    if (props) {
+      return train.sgd(props.learningRate);
+    } else {
+      throw new TypeError('Illegal properties to init optimizer');
```
Comment (FeelyChau) [inline src/utils/optimizer-types.ts:58]: The non-null assert can be moved to outside of the `switch`.

File: webpack.config.js [hunk 9]
```diff
@@ -1,19 +1,24 @@
...
-
+const BundleAnalyzerPlugin = require('webpack-bundle-analyzer').BundleAnalyzerPlugin;
 const createConfig = (target) => {
   return {
     mode: "production",
     devtool: "source-map",
     context: path.resolve(__dirname),
     entry: {
-      index: './dist/index.js'
+      //index: './dist/index.js',
```
Comment (FeelyChau) [inline webpack.config.js:9]: should enable the index entry.

**Review by FeelyChau — CHANGES_REQUESTED** [review]

File: package.json (hunk 3 above)
Comment (yorkie) [inline package.json:13]: This should not be in dependencies.

File: package.json [hunk 10]
```diff
@@ -4,10 +4,13 @@
   "description": "A JavaScript library for feature engineering on datasets",
   "main": "dist/index.js",
   "dependencies": {
+    "@babel/preset-env": "^7.15.4",
     "@tensorflow/tfjs-core": "^3.8.0",
     "@tensorflow/tfjs-layers": "^3.8.0",
+    "babel-loader": "^8.2.2",
```
Comment (yorkie) [inline package.json:10]: Ditto

File: package.json [hunk 11]
```diff
@@ -4,10 +4,13 @@
   "description": "A JavaScript library for feature engineering on datasets",
   "main": "dist/index.js",
   "dependencies": {
+    "@babel/preset-env": "^7.15.4",
```
Comment (yorkie) [inline package.json:7]: Ditto.

File: src/model/linear-model/logistic-regression-predictor.ts [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+  }
+
+  /**
+   * Load model from dumped model json string.
+   * @param modelJson Dumped model json string
+   * @returns model itself
+   */
+  public async fromJson(modelJson: string): Promise<LogisticRegressionPredictor> {
+    const modelParams = JSON.parse(modelJson);
+    if (modelParams.name !== 'LogisticRegression'){
```
Comment (FeelyChau) [inline src/model/linear-model/logistic-regression-predictor.ts:50]:
```suggestion
    if (modelParams.name !== 'LogisticRegression') {
```

File: src/model/linear-model/logistic-regression-predictor.ts [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+      throw new RangeError(`${modelParams.name} is not Logistic Regression`);
+    }
+    const { classes, modelWeights, featureSize } = modelParams;
+    this.featureSize = featureSize;
+    classes && this.initClasses(classes, 'binary-only');
```
Comment (FeelyChau) [inline src/model/linear-model/logistic-regression-predictor.ts:55]:
```suggestion
    if (classes) {
        this.initClasses(classes, 'binary-only');
    }
```

**Review by FeelyChau — CHANGES_REQUESTED** [review]

File: src/model/linear-model/linear-regression.ts [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+   * `optimizerProps`: parameters used to init corresponding optimizer, you can refer to documentations in tensorflow.js
+   *  (https://js.tensorflow.org/api/latest/#Training-Optimizers) to find the supported initailization paratemters for a
+   *  given type of optimizer. For example, `{ learningRate: 0.1, beta1: 0.1, beta2: 0.2, epsilon: 0.1 }` could be used
+   *  to initialize adam optimizer.
+   */
+  constructor(params: LinearRegressionParams = {}) {
+    super();
+    this.fitIntercept = params.fitIntercept !== false;
+    this.normalize = params.normalize;
+    //this.optimizer = params.optimizer && params.optimizer instanceof Optimizer ? params.optimizer : train.adam(0.1);
```
Comment (FeelyChau) [inline src/model/linear-model/linear-regression.ts:82]: Remove unused lines.

File: src/model/linear-model/logistic-regression.ts [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+   *  - 'adam': adam optimizer
+   *  - 'adamax': adamax optimizer
+   *  - 'rmsprop': rmsprop optimizer
+   *
+   * `optimizerProps`: parameters used to init corresponding optimizer, you can refer to documentations in tensorflow.js
+   *  (https://js.tensorflow.org/api/latest/#Training-Optimizers) to find the supported initailization paratemters for a
+   *  given type of optimizer. For example, `{ learningRate: 0.1, beta1: 0.1, beta2: 0.2, epsilon: 0.1 }` could be used
+   *  to initialize adam optimizer.
+   */
+  constructor(params: LogisticRegressionParams = {
```
Comment (FeelyChau) [inline src/model/linear-model/logistic-regression.ts:90]: If pass `{ c: 1 }`, the penalty will be set to `undefined`, is this expected?

File: src/model/linear-model/logistic-regression.ts [hunk 16]
```diff
@@ -0,0 +1,281 @@
...
+  private initModel(inputShape: number, outputShape: number, useBias = true, modelWeights: Tensor[] = []): void {
+    const model = sequential();
+    const c = this.c;
+    const penalty = this.penalty;
+    model.add(layers.dense({
+      inputShape: [ inputShape ],
+      units: outputShape,
+      useBias,
+      activation: 'sigmoid',
+      kernelRegularizer: penalty == 'l2' ? regularizers.l2({ l2: c }) : penalty == 'l1' ? regularizers.l1({ l1: c }) : null }));
```
Comment (FeelyChau) [inline src/model/linear-model/logistic-regression.ts:115]:
```suggestion
      kernelRegularizer: penalty === 'l2' ? regularizers.l2({ l2: c }) : penalty === 'l1' ? regularizers.l1({ l1: c }) : null }));
```

File: src/model/linear-model/logistic-regression.ts [hunk 17]
```diff
@@ -0,0 +1,281 @@
...
+    model.compile({
+      optimizer: this.optimizer,
+      loss: losses.sigmoidCrossEntropy
+    });
+    modelWeights.length && model.setWeights(modelWeights);
```
Comment (FeelyChau) [inline src/model/linear-model/logistic-regression.ts:120]:
```suggestion
  if (modelWeights.length > 0) {
    model.setWeights(modelWeights);
  }
```

File: src/model/linear-model/logistic-regression.ts [hunk 18]
```diff
@@ -0,0 +1,281 @@
...
+    const { x, y } = this.validateData(xData, yData);
+    const nFeature = x.shape[1];
+    if (!this.model) {
+      await this.initClasses(y, 'binary-only');
+      const outputShape = this.isBinaryClassification() ? 1 : this.classes().shape[0];
+      this.initModel(nFeature, outputShape, this.fitIntercept);
+      this.featureSize = nFeature;
+      this.outputSize = outputShape;
+    } else {
+      if (nFeature != this.featureSize) {
```
Comment (FeelyChau) [inline src/model/linear-model/logistic-regression.ts:139]:
```suggestion
      if (nFeature !== this.featureSize) {
```

File: src/model/linear-model/logistic-regression.ts [hunk 19]
```diff
@@ -0,0 +1,281 @@
...
+    const x = checkArray(xData, 'float32');
+    const scores = this.model.predict(x);
+    if (scores instanceof Array){
+      return stack(scores);
+    } else {
+      return scores;
+    }
+  }
+
+  public getCoef(): { 'coefficients': Tensor, 'intercept': Tensor} {
```
Comment (FeelyChau) [inline src/model/linear-model/logistic-regression.ts:221]:
```suggestion
  public getCoef(): { coefficients: Tensor, intercept: Tensor } {
```

**Review by FeelyChau — CHANGES_REQUESTED** [review]

**Comment by FeelyChau** [comment]
LGTM.

**Review by FeelyChau — APPROVED** [review]

**Pull request merged**